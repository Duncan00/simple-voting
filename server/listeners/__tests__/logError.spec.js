'use strict';

const logError = require('../logError');

describe('logError', () => {
	const mockLogger = {
		error: jest.fn()
	};

	const ctx = {
		get: id => id,
		method: 'method',
		length: 'length',
		href: 'href',
		state: {
			request_cycle_trace_id: 'request_cycle_trace_id',
			response_time: 'response_time',
			user: {
				id: 'id'
			}
		},
		ip: 'ip',
		status: 'status-from-ctx'
	};

	beforeEach(() => {
		mockLogger.error.mockReset();
	});
	test('ctx status missing -> use err.status', () => {
		const error = new Error('testing');
		error.stack = undefined;
		error.message = 'message in error.message';
		error.status = 'status-from-error';
		logError({logger: mockLogger})(error, {
			...ctx,
			status: undefined
		});

		expect(mockLogger.error.mock.calls).toMatchSnapshot(
			'message from message'
		);
	});
	test('should match snapshot when call logger.error(), error.stack', () => {
		const error = new Error('testing');
		error.stack = 'mock-stack';
		logError({logger: mockLogger})(error, ctx);

		expect(mockLogger.error.mock.calls).toMatchSnapshot();
	});
	test('error without stack', () => {
		const error = new Error('testing');
		error.stack = undefined;
		error.message = 'message in error.message';
		logError({logger: mockLogger})(error, ctx);

		expect(mockLogger.error.mock.calls).toMatchSnapshot(
			'message from message'
		);
	});
	test('error without stack and message', () => {
		const error = new Error('testing');
		error.stack = undefined;
		error.message = undefined;
		error.details = 'message in details';
		logError({logger: mockLogger})(error, ctx);

		expect(mockLogger.error.mock.calls).toMatchSnapshot(
			'message from details'
		);
	});
	test('ctx with user id', () => {
		const error = new Error('testing');
		error.stack = undefined;
		error.message = undefined;
		error.details = 'message in details';
		logError({logger: mockLogger})(
			error,
			Object.assign({}, ctx, {
				state: {
					user: {
						id: 'user-id'
					}
				}
			})
		);

		expect(mockLogger.error.mock.calls).toMatchSnapshot();
	});
});
