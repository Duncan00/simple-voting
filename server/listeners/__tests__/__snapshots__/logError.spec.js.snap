// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logError ctx status missing -> use err.status: message from message 1`] = `
Array [
  Array [
    Object {
      "MONITOR_EVENT_KEY": "unexpected-internal-error",
      "details": "",
      "httpRequest": Object {
        "latency": "response_time",
        "referer": "referer",
        "remoteIp": "ip",
        "requestMethod": "method",
        "requestUrl": "href",
        "responseSize": "length",
        "status": "status-from-error",
        "userAgent": "user-agent",
      },
      "message": "message in error.message",
      "request_cycle_trace_id": "request_cycle_trace_id",
      "user": "id",
    },
  ],
]
`;

exports[`logError ctx with user id 1`] = `
Array [
  Array [
    Object {
      "MONITOR_EVENT_KEY": "unexpected-internal-error",
      "details": "message in details",
      "httpRequest": Object {
        "latency": undefined,
        "referer": "referer",
        "remoteIp": "ip",
        "requestMethod": "method",
        "requestUrl": "href",
        "responseSize": "length",
        "status": "status-from-ctx",
        "userAgent": "user-agent",
      },
      "message": "message in details",
      "request_cycle_trace_id": undefined,
      "user": "user-id",
    },
  ],
]
`;

exports[`logError error without stack and message: message from details 1`] = `
Array [
  Array [
    Object {
      "MONITOR_EVENT_KEY": "unexpected-internal-error",
      "details": "message in details",
      "httpRequest": Object {
        "latency": "response_time",
        "referer": "referer",
        "remoteIp": "ip",
        "requestMethod": "method",
        "requestUrl": "href",
        "responseSize": "length",
        "status": "status-from-ctx",
        "userAgent": "user-agent",
      },
      "message": "message in details",
      "request_cycle_trace_id": "request_cycle_trace_id",
      "user": "id",
    },
  ],
]
`;

exports[`logError error without stack: message from message 1`] = `
Array [
  Array [
    Object {
      "MONITOR_EVENT_KEY": "unexpected-internal-error",
      "details": "",
      "httpRequest": Object {
        "latency": "response_time",
        "referer": "referer",
        "remoteIp": "ip",
        "requestMethod": "method",
        "requestUrl": "href",
        "responseSize": "length",
        "status": "status-from-ctx",
        "userAgent": "user-agent",
      },
      "message": "message in error.message",
      "request_cycle_trace_id": "request_cycle_trace_id",
      "user": "id",
    },
  ],
]
`;

exports[`logError should match snapshot when call logger.error(), error.stack 1`] = `
Array [
  Array [
    Object {
      "MONITOR_EVENT_KEY": "unexpected-internal-error",
      "details": "",
      "httpRequest": Object {
        "latency": "response_time",
        "referer": "referer",
        "remoteIp": "ip",
        "requestMethod": "method",
        "requestUrl": "href",
        "responseSize": "length",
        "status": "status-from-ctx",
        "userAgent": "user-agent",
      },
      "message": "mock-stack",
      "request_cycle_trace_id": "request_cycle_trace_id",
      "user": "id",
    },
  ],
]
`;
